sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/DefaultSynchronousMetricStorage.java

```diff
@@ -146,13 +158,25 @@ public MetricData collect(
             ? registeredReader.getLastCollectEpochNanos()
             : startEpochNanos;
 
+    ConcurrentHashMap<Attributes, AggregatorHandle<T, U>> aggregatorHandles;
+    if (reset) {
+      long stamp = sl.writeLock();
+      try {
+        aggregatorHandles = this.aggregatorHandles;
+        this.aggregatorHandles = new ConcurrentHashMap<>();
+      } finally {
+        sl.unlockWrite(stamp);
+      }
+    } else {
+      aggregatorHandles = this.aggregatorHandles;
+    }
+
     // Grab aggregated points.
     List<T> points = new ArrayList<>(aggregatorHandles.size());
     aggregatorHandles.forEach(
         (attributes, handle) -> {
           T point = handle.aggregateThenMaybeReset(start, epochNanos, attributes, reset);
           if (reset) {
-            aggregatorHandles.remove(attributes, handle);
```

Comment:
Here's the bug:
- On line 153 we grab the aggregated data from the handle
- On line 155 we remove the handle from the map
- Removing the handle from the map prompts new measurements for the attributes to add a new handle to the map
- Aggregating the data from the handle and removing it from the map is not atomic. This results in the possibility of lost writes: where another recording thread records to the handle after it has been aggregated but before it is removed from the map.

The solution is to guard the `aggregatorHandles` map with a read write lock. When collecting, we obtain a write lock and replace the `aggregatorHandles` map with a new instance. When recording, we obtain a read lock, which is cheap and blocks the collect thread from replacing the `aggregatorHandles` map until the record thread releases the read lock.